### PR TITLE
change the method type of generate_square_subsequent_mask

### PR DESCRIPTION
**Pull request by Kyeongpil**
Change the method type of generate_square_subsequent_mask to static method in order to be accessed by
`nn.Transformer.generate_square_subsequent_mask(sz)`

instead of
`model = nn.Transformer(args)`
`model.generate_square_subsequent_mask(sz)`

By changing this method as a static method, we can directly access/use the method without creating an instance.

### CODE REVIEW

**Comment by rudvlf0413** [comment]
There has been no review process for this pull request.
@apaszke, Could you review this pull request?

**Comment by soumith** [comment]
@zhangguanheng66 please review

**Comment by zhangguanheng66** [comment]
@rudvlf0413 it's not necessary to be within nn.Transform module. You could move `generate_square_subsequent_mask` to torch.nn.functional file.  Or we could even drop the `generate_square_subsequent_mask` func since it's not part of the transformer model.

**Review by zhangguanheng66 — COMMENTED** [review]
@rudvlf0413 it's not necessary to be within nn.Transform module. You could move generate_square_subsequent_mask to torch.nn.functional file. Or we could even drop the generate_square_subsequent_mask func since it's not part of the transformer model.

**Comment by rudvlf0413** [comment]
@zhangguanheng66 I totally agree with your opinion.
I move the function to the torch.nn.functional.
Could you review the codes?

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -121,21 +122,21 @@ def forward(self, src, tgt, src_mask=None, tgt_mask=None,
...
-        return mask
-
     def _reset_parameters(self):
         r"""Initiate parameters in the transformer model."""
 
         for p in self.parameters():
             if p.dim() > 1:
                 xavier_uniform_(p)
 
+    def generate_square_subsequent_mask(self, sz):
```
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: IMO, we should eventually retire this function since it's not relevant to any components in `nn.Transformer`. You could add a warning message for this func and we could retire it after two releases. @cpuhrsch

**Review by zhangguanheng66 — CHANGES_REQUESTED** [review]

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: Yes that sounds like a reasonable approach. However, this function is simple enough that I'm not sure it warrants being a separate entry in our nn.functional module given the sort of other functions it would neighbor (e.g. conv2d).

@rudvlf0413 is there an advantage to having this be its own functional or static method?

Comment (rudvlf0413) [reply torch/nn/modules/transformer.py]: @cpuhrsch We can use the function without creating the Transformer object if we separate the function from the Transformer module.
This is helpful when we custom Transformer module or order modules that require that function.

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: @rudvlf0413 given the code is very short, would your usecase be served by copy-pasting the code?

Comment (rudvlf0413) [reply torch/nn/modules/transformer.py]: @cpuhrsch Yeah, Although the function is separated from the main module, the body of the function is the same. However, we will feel convenient if we can use the function directly.

Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: @cpuhrsch I feel it's really a simple functional and is not necessary to keep it around.  @rudvlf0413 could you give us some usage cases, which are not related to `nn.Transformer` modules. :)

Comment (rudvlf0413) [reply torch/nn/modules/transformer.py]: @cpuhrsch @zhangguanheng66 I think this function can be applied to other seq-to-seq models such as "Convolutional sequence to sequence learning (https://arxiv.org/abs/1705.03122)" that requires masking operations when generating target sequences in training session.

I found a similar function from FairSeq as below
https://github.com/pytorch/fairseq/blob/master/fairseq/models/transformer.py#L593

Although the transformer module only uses generate_square_subsequent_mask currently, I think some modules will need this function in the future. Thus, I think it is better to locate the function in nn.functional.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Thank you for your contribution!

Given the brevity and specificity of the function and given that the fact that it's not used within Transformers, please copy-paste the code for your purposes. We'll then revisit this if later on there are requests for this kind of function at an nn.functional level.

For future reference, please also avoid updating submodules as part of PRs.
